src/models.jl

```diff
@@ -169,5 +171,129 @@ function _l2norm_vec!(out, Vs::Tuple)
+function preallocate_solve_ROF_ADMM(::Type{T}, img) where T
+    # Use similar to allow construct CuArray when `img` is a CuArray
+
+    # split variable for ∇
+    d = ntuple(i->fill!(similar(img, T), zero(T)), ndims(img))
+    # dual variable
+    b = map(copy, d)
+
+    # precomputed the negative Laplacian kernel in frequency space
+    DTD = _negative_flaplacian_freqkernel(img)
+
+    # buffer for u subproblem
+    #   minᵤ 0.5μ||u-f||₂² + 0.5∑ᵢ(λ||dᵢ - ∇ᵢu - bᵢ||₂²)
+    # For generality to arbitrary dimension, fft-based solver is applied
+    #   u = real.(ifft(fft(RHS)./fft(LHS)))
+    # where LHS is actually a convolution operator thus we can apply convolution theorem.
+    Δbd = similar(first(b))
+    LHS = similar(first(d)) # left-hand side: μI + Δ
+    RHS = similar(LHS) # right-hand side: μf + λ∑ᵢ(∇ᵢ'(dᵢ - bᵢ))
+    fft_tmp = similar(LHS, Complex{eltype(T)})
+
+    return d, b, DTD, Δbd, LHS, RHS, fft_tmp
+end
+
+"""
+    solve_ROF_ADMM!(out, buffer, img, λ, μ; kwargs...)
+
+The in-place version of [`solve_ROF_ADMM`](@ref).
+"""
+function solve_ROF_ADMM!(out::AbstractArray{T}, buffer, img, λ, μ; kwargs...) where T
+    # seperate a stub method to reduce latency
+    FT = float32(T)
+    if FT == T
+        _solve_ROF_ADMM_anisotropic!(out, buffer, img, Float32(λ), Float32(μ); kwargs...)
+    else
+        _solve_ROF_ADMM_anisotropic!(out, buffer, FT.(img), Float32(λ), Float32(μ); kwargs...)
+    end
+end
+
+function _solve_ROF_ADMM_anisotropic!(
+        out::AbstractArray,
+        (d, b, DTD, Δbd, LHS, RHS, fft_tmp)::Tuple,
+        img::AbstractArray,
+        λ::Float32, μ::Float32; num_iters::Integer=200)
+    # The notation and algorithm follows reference [1] with one modification:
+    u, f = out, img
+
+    # u-subproblem is solved using direct fft-based method instead of using Gauss–Seidel
+    # method. This is because Gauss–Seidel method itself is a solver for 2-dimensional case,
+    # while FFT-based solver support arbitrary dimensional. However, we might still want to
+    # provide a specialization solver for the 2D case using Gauss-Seidel method and see if
+    # it helps speaking of performance.
+
+    # soft-thresholding: d-subproblem is a simplified lasso problem
+    S(x, λ) = sign(x) * max(abs(x)-λ, zero(x))
+
+    # initialization
+    foreach(x->fill!(x, zero(eltype(x))), d)
+    foreach(x->fill!(x, zero(eltype(x))), b)
+    @. LHS = λ + μ * DTD
+    # apply the algorithm below Eq. (4.2)
+    for _ in 1:num_iters
+        @. RHS = λ/μ * f
+        for i in 1:length(d)
+            # negative adjoint gradient is the reverse finite difference
+            @. Δbd = b[i] - d[i]
+            # reuse d here as buffer as it will be updated later in d-subproblem
+            RHS .+= fdiff!(d[i], Δbd; dims=i, rev=true)
+        end
+        # TODO(johnnychen94): optimize the memory allocation
+        fft_tmp .= μ .* fft(RHS)./LHS
+        u .= real.(ifft(fft_tmp))
```

Comment:
This is the performance hot spot:

<img src="https://user-images.githubusercontent.com/8684355/140610994-29714087-140e-40d5-8c05-bede909b77a9.png" alt="preview.png" width="600">

I guess we can use in-place version `fft!` or real version `rfft` to boost it.